### PR TITLE
Gradle plugin

### PR DESCRIPTION
**Pull request by dmdashenkov**
In this PR we introduce the ProtoData Gradle plugin.

The plugin adds the necessary tasks and config for launching ProtoData.

## Configuration

To use ProtoData, apply and configure the plugin as follows:

```kotlin

plugins {
    id("io.spine.proto-data") version("<ProtoData version>")
    
    // Also must apply:
    java
    id("com.google.protobuf") version("...")
}

protoData {
    plugins("com.acme.ValidationPlugin", "com.acme.ColumnsPlugin", "org.example.EtcPlugin")
    renderers("com.acme.InsertionPointPrinter", "com.acme.JavaValidationRenderer", "org.example.EtcRenderer")
    optionProviders("org.example.MyOptionsProvider")
}
```

There can be more configuration if necessary. See `io.spine.protodata.gradle.Extension` for more.

## Installation

ProtoData is an executable installed onto a computer.

Users can install ProtoData using the Gradle plugin by running:
```
./gradlew installProtoData
```

To change the installation directory, users may specify a Gradle property: 
```
./gradlew installProtoData -PprotoDataLocation="/path/to/target/dir"
```

## Launching

ProtoData is tied to the project build and will launch automatically in Java projects.

To launch ProtData manually, run 
```
./gradlew launchProtoData
```

If a ProtoData executable is not in the `PATH` environmental variable at this point, the task will fail.
To bypass `PATH` and run ProtoData from a known location on disc, users should, once again, submit the `protoDataLocation` property:
```
./gradlew launchProtoData -PprotoDataLocation="/path/to/target/dir"
```

This applies to any build which invokes the `launchProtoData` task.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: gradle-plugin/src/main/kotlin/io/spine/protodata/gradle/Plugin.kt [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+ */
+@Suppress("UnstableApiUsage") // Gradle Property API.
+public class Extension(private val project: Project) {
+
+    internal  val plugins: ListProperty<String> =
+        project.objects.listProperty(String::class.java)
+
+    /**
+     * Passes given names of Java classes to ProtoData as the `io.spine.protodata.plugin.Plugin`
+     * classes.
```
Comment (armiol) [inline gradle-plugin/src/main/kotlin/io/spine/protodata/gradle/Plugin.kt:99]: Let's save this lonely word from being an orphan for the rest of its days.

File: gradle-plugin/src/main/kotlin/io/spine/protodata/gradle/Plugin.kt [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    public fun plugins(vararg classNames: String) {
+        plugins.addAll(classNames.toList())
+    }
+
+    internal val renderers: ListProperty<String> =
+        project.objects.listProperty(String::class.java)
+
+    /**
+     * Passes given names of Java classes to ProtoData as the `io.spine.protodata.renderer.Renderer`
+     * classes.
```
Comment (armiol) [inline gradle-plugin/src/main/kotlin/io/spine/protodata/gradle/Plugin.kt:110]: Same here.

File: gradle-plugin/src/main/kotlin/io/spine/protodata/gradle/Plugin.kt [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+        project.objects.directoryProperty().convention(
+            project.layout.projectDirectory.dir("generated/main/java")
+        )
+
+    /**
+     * The location where the sources processed by ProtoData can be found.
+     *
+     * By default, `generated/main/java`.
+     */
+    public var source: Any
```
Comment (armiol) [inline gradle-plugin/src/main/kotlin/io/spine/protodata/gradle/Plugin.kt:154]: Judging on the description, it's rather a destination.

Did you mean "the sources to process by ProtoData"?

File: gradle-plugin/src/main/kotlin/io/spine/protodata/gradle/Plugin.kt [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+    }
+
+    internal val optionProviders: ListProperty<String> =
+        project.objects.listProperty(String::class.java)
+
+    /**
+     * Passes given names of Java classes to ProtoData as
+     * the `io.spine.protodata.option.OptionsProvider` classes.
+     */
+    public fun optionProviders(vararg classNames: String) {
```
Comment (armiol) [inline gradle-plugin/src/main/kotlin/io/spine/protodata/gradle/Plugin.kt:123]: Can we have an issue to implement an alternative syntax as discussed?
```
options("this/is/where/my/options.proto", ... )
```

File: gradle-plugin/src/main/kotlin/io/spine/protodata/gradle/Plugin.kt [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+import org.gradle.api.tasks.Copy
+import org.gradle.api.tasks.Exec
+import org.gradle.api.tasks.compile.JavaCompile
+import org.gradle.kotlin.dsl.withType
+import org.gradle.api.Plugin as GradlePlugin
+
+/**
+ * The ProtoData Gradle plugin.
+ *
+ * Adds the `installProtoData` task which installs the ProtoData executable onto
```
Comment (armiol) [inline gradle-plugin/src/main/kotlin/io/spine/protodata/gradle/Plugin.kt:47]: Can we please say more on why one ever needs just to `installProtoData`, and not to `launch...` it?

I think that `launchProtoData` is much more important, and will be used in 99% of cases. Therefore, I would mention `installProtoData` later, and in a context of a potential usage scenario.

**Review by armiol — CHANGES_REQUESTED** [review]
@dmdashenkov LGTM in general. Please see some comments.

File: gradle-plugin/src/main/kotlin/io/spine/protodata/gradle/Plugin.kt (hunk 4 above)
Comment (dmdashenkov) [reply gradle-plugin/src/main/kotlin/io/spine/protodata/gradle/Plugin.kt:123]: Done. See https://github.com/SpineEventEngine/ProtoData/issues/19.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

**Review by armiol — APPROVED** [review]

**Pull request merged**